### PR TITLE
Feat/apple native ocr

### LINKED ISSUE #805: [bounty] implement apple native OCR in pure rust
/bounty 250 

definition of done
- apple native OCR in pure rust
- no memory leaks
- no worse performance or accuracy

maybe using https://github.com/yury/cidre

### PR DESCRIPTION
**Pull request by b4s36t4**
## description
This PR used `cidre` to implement apple native OCR instead of using swift based binary to perform OCR.

We temporarily create the image to perform OCR using `cidre` and delete the file again.

/claim #805
/closes #805

related issue: #

## type of change
- [ ] bug fix
- [x] new feature
- [ ] breaking change
- [ ] documentation update

if relevant add screenshots or screen captures to prove that this PR works to save us time.

https://github.com/user-attachments/assets/9a2b87c3-0a77-4bd1-87a1-f28f6f80b1bd

## checklist
- [ ] MOST IMPORTANT: this PR will require less than 30 min to review, merge, and release to production and not crash in the hand of thousands of users
- [x] i have read the [CONTRIBUTING.md](https://github.com/mediar-ai/screenpipe/blob/main/CONTRIBUTING.md) file 
- [ ] i have updated the documentation if necessary
- [ ] my changes generate no new warnings
- [ ] i have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: screenpipe-vision/src/apple.rs [hunk 1]
```diff
@@ -1,86 +1,135 @@
...
+    ocr_result: String,
+    text_elements: Vec<OcrTextElement>,
+    overall_confidence: f32,
 }
+
 #[cfg(target_os = "macos")]
 pub fn perform_ocr_apple(
     image: &DynamicImage,
-    languages: Vec<screenpipe_core::Language>,
+    _languages: Vec<screenpipe_core::Language>,
```
Comment (louis030195) [inline screenpipe-vision/src/apple.rs:32]: we need the language feature

**Comment by louis030195** [comment]
nice !

any chance to share any diff in performance and if it creates any memory leak? can use the `leak` cli (ask AI)

@b4s36t4

**Comment by louis030195** [comment]
```

   Compiling enigo v0.1.3
   Compiling strsim v0.10.0
   Compiling sqlx-macros v0.7.2
error[E0658]: referencing statics in constant functions is unstable
  --> /Users/louisbeaumont/.cargo/git/checkouts/cidre-e660120dbd7e907a/62b50e8/cidre/src/cm/time/range.rs:40:18
   |
40 |         unsafe { kCMTimeRangeZero }
   |                  ^^^^^^^^^^^^^^^^
   |
   = note: see issue #119618 <https://github.com/rust-lang/rust/issues/119618> for more information
   = note: `static` and `const` variables can refer to other `const` variables. A `const` variable, however, cannot refer to a `static` variable.
   = help: to fix this, the value can be extracted to a `const` and then used.

error[E0658]: referencing statics in constant functions is unstable
  --> /Users/louisbeaumont/.cargo/git/checkouts/cidre-e660120dbd7e907a/62b50e8/cidre/src/cm/time/range.rs:46:18
   |
46 |         unsafe { kCMTimeRangeInvalid }
   |                  ^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #119618 <https://github.com/rust-lang/rust/issues/119618> for more information
   = note: `static` and `const` variables can refer to other `const` variables. A `const` variable, however, cannot refer to a `static` variable.
   = help: to fix this, the value can be extracted to a `const` and then used.

error[E0658]: referencing statics in constant functions is unstable
   --> /Users/louisbeaumont/.cargo/git/checkouts/cidre-e660120dbd7e907a/62b50e8/cidre/src/cm/time.rs:291:18
    |
291 |         unsafe { kCMTimeZero }
    |                  ^^^^^^^^^^^
    |
    = note: see issue #119618 <https://github.com/rust-lang/rust/issues/119618> for more information
    = note: `static` and `const` variables can refer to other `const` variables. A `const` variable, however, cannot refer to a `static` variable.
    = help: to fix this, the value can be extracted to a `const` and then used.

   Compiling cocoa v0.22.0
   Compiling opentelemetry-proto v0.4.0
   Compiling core-graphics v0.19.2
   Compiling opentelemetry-http v0.10.0
For more information about this error, try `rustc --explain E0658`.
   Compiling opentelemetry-semantic-conventions v0.13.0
error: could not compile `cidre` (lib) due to 3 previous errors
warning: build failed, waiting for other jobs to finish...
(base) louisbeaumont@MacBook-Pro-2:~/Documents/screenpipe$ cargo build --release --features metal 
```

does not compile on my side

File: screenpipe-vision/src/apple.rs (hunk 1 above)
Comment (b4s36t4) [reply screenpipe-vision/src/apple.rs:32]: Cidre doesn't allow langauge selection. That's why I had used revision.

**Comment by m13v** [comment]
soe me:

(base) matthewdi@MacBook-Pro screen-pipe % cargo build --release --features metal
   Compiling libc v0.2.164
   Compiling cidre v0.4.0 (https://github.com/yury/cidre.git?rev=f05c428#f05c4288)
   Compiling cidre v0.5.0 (https://github.com/yury/cidre.git#62b50e8b)
error: failed to run custom build command for `cidre v0.4.0 (https://github.com/yury/cidre.git?rev=f05c428#f05c4288)`

Caused by:
  process didn't exit successfully: `/Users/matthewdi/Desktop/screenpipe/screen-pipe/target/release/build/cidre-2d9a73e8726b5a24/build-script-build` (exit status: 101)
  --- stdout
  cargo:rustc-link-lib=clang_rt.osx
  cargo:rustc-link-search=/Library/Developer/CommandLineTools/usr/lib/clang/16/lib/darwin
  cargo:rustc-link-search=framework=/System/Library/PrivateFrameworks
  cargo:rustc-link-search=framework=/Library/Apple/System/Library/PrivateFrameworks
  cargo:rustc-link-search=native=/Users/matthewdi/Desktop/screenpipe/screen-pipe/target/release/build/cidre-3df686e32bd43c22/out/Release

  --- stderr
  xcode-select: error: tool 'xcodebuild' requires Xcode, but active developer directory '/Library/Developer/CommandLineTools' is a command line tools instance
  thread 'main' panicked at /Users/matthewdi/.cargo/git/checkouts/cidre-e660120dbd7e907a/f05c428/cidre/build.rs:145:5:
  assertion failed: status.success()
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
error: failed to run custom build command for `cidre v0.5.0 (https://github.com/yury/cidre.git#62b50e8b)`

Caused by:
  process didn't exit successfully: `/Users/matthewdi/Desktop/screenpipe/screen-pipe/target/release/build/cidre-35b8ed00e8349e9a/build-script-build` (exit status: 101)
  --- stdout
  cargo:rustc-link-lib=clang_rt.osx
  cargo:rustc-link-search=/Library/Developer/CommandLineTools/usr/lib/clang/16/lib/darwin
  cargo:rustc-link-search=framework=/System/Library/PrivateFrameworks
  cargo:rustc-link-search=framework=/Library/Apple/System/Library/PrivateFrameworks
  cargo:rustc-link-search=native=/Users/matthewdi/Desktop/screenpipe/screen-pipe/target/release/build/cidre-36d1721d1528b9f3/out/Release

  --- stderr
  xcode-select: error: tool 'xcodebuild' requires Xcode, but active developer directory '/Library/Developer/CommandLineTools' is a command line tools instance
  thread 'main' panicked at /Users/matthewdi/.cargo/git/checkouts/cidre-e660120dbd7e907a/62b50e8/cidre/build.rs:145:5:
  assertion failed: status.success()
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
(base) matthewdi@MacBook-Pro screen-pipe %

**Comment by b4s36t4** [comment]
@m13v I believe the error you are facing is not related to this PR, it seems `cidre` is dependent on `xcode` to build. I suggest please install `xcode` and try again to build. Please let me know if that's not the case.

**Comment by louis030195** [comment]
not going to test again unless there is this:

>any chance to share any diff in performance and if it creates any memory leak? can use the leak cli (ask AI)

**Comment by b4s36t4** [comment]
@louis030195 I'm trying to do so, but I'm not able to. 

`  error: leak sanitizer is not supported for this target` getting this error when running `  error: leak sanitizer is not supported for this target`

Even this `RUSTFLAGS="-Z sanitizer=address" cargo run --bin screenpipe` command is also giving error. Can you help me on how to test(leak test)?

**Comment by louis030195** [comment]
tried again, updating from rust 1.81 to 1.83 fixed the issue, however:

i get spikes of cpu from 15 to 140% every 2 seconds

some reasons:

- temporary File I/O: Every OCR operation writes to disk and reads back
- inefficient String Handling: Concatenating strings with `+=` in a loop
- Unnecessary JSON Serialization/Deserialization

**Review by ologbonowiwi — COMMENTED** [review]
I see we have a bunch of unwraps on this impl, I suggest we handle these before merging. Otherwise app will panic, probably we could use a few defaults or something when the result is err

**Comment by louis030195** [comment]
https://github.com/yury/cidre/commit/b6b685ca0c84ed7f8c8344503561fe44adf68eef

also now `cidre` supports language @b4s36t4

**Comment by louis030195** [comment]
fyi if no news im going to do this myself

**Comment by louis030195** [comment]
any news?

**Comment by b4s36t4** [comment]
Sorry, but been facing some issue. As you've mentioned writing images into disk is causing a cpu spike which I want to tackle.  But I'm getting this error when trying to implement ocr with `initWithData:options` function (update cidre in local to support the function)

error:
```
"CRImage Reader Detector was given zero-dimensioned image (0 x 0)"
```

**Comment by ologbonowiwi** [comment]
@b4s36t4 idk if its possible, but we could use a buffer to write multiple times into the disk at once. Like adding N images, when it gets to limit X we write X images at once. Maybe this could solve the issue?

**Comment by ologbonowiwi** [comment]
@b4s36t4 idk if its possible, but we could use a buffer to write multiple times into the disk at once. Like adding N images, when it gets to limit X we write X images at once. Maybe this could solve the issue?

**Comment by louis030195** [comment]
> @b4s36t4 idk if its possible, but we could use a buffer to write multiple times into the disk at once. Like adding N images, when it gets to limit X we write X images at once. Maybe this could solve the issue?

hmm no

#811 we want to do real time streaming of frames with OCR so any added latency is not great

**Comment by b4s36t4** [comment]
@louis030195 I'm still not able to run `leaks` cli against the screenpipe.

first command that you shared is giving `cargo` command is not found.
as for the second it's  getting this error ` Job 1, 'leaks target/release/screenpipe…' terminated by signal SIGSEGV (Address boundary error)`

**Comment by b4s36t4** [comment]
Also ran the benchmark for the updated implementation, not sure weather it's high or low to current benchmarks. Please checkout and let me know.

> I'm using Apple M1 Pro.

[banch.txt](https://github.com/user-attachments/files/18054411/banch.txt)

**Comment by b4s36t4** [comment]
Raised a PR to `cidre` for performing ocr with bytes instead of saving image.

https://github.com/yury/cidre/pull/30

File: screenpipe-vision/src/apple.rs (hunk 1 above)
Comment (b4s36t4) [reply screenpipe-vision/src/apple.rs:37]: Will be adding languages support by tmrw (lost the code when switching branches).

**Comment by b4s36t4** [comment]
@louis030195 I ran the benchmark again on `main` still I'm seeing `0.68` as average accuracy. Maybe it's our macOS version difference. I'm on `Sonama`, what were you using!?

**Comment by louis030195** [comment]
@b4s36t4 sequoia 15.1

**Comment by louis030195** [comment]
Screenshot 2024-12-09 at 2 14 44 PM

**Comment by b4s36t4** [comment]
I think it does depend on the OS Version (hypothetically ocr has been improved with macOS upgrade ;))

**Comment by b4s36t4** [comment]
@louis030195 Let me know if I'm missing anything, would love to tackle and close this :)

**Comment by louis030195** [comment]
> @louis030195 Let me know if I'm missing anything, would love to tackle and close this :)

CPU usage is still worse than on main, maybe try grayscale, not iterating/building the language list at every iteration or other stuff

**Comment by louis030195** [comment]
increased bounty to 250

**Comment by b4s36t4** [comment]
@louis030195 I think I was able to reduce the CPU usage. Monitored the cpu usage specifically `ocr_benchmark` without changes it was around 150, 200% usage. But after the changes it was around `90` like that, even accuracy also increased a bit (not sure why).

But there seems to be an issue with bench, if I put `request.set_uses_lang_correction(true);` benchmark was able to complete within the 10sec timeframe but this is causing cpu spike, if I put `false` there, benchmark is taking around `18-19s` to complete but the cpu usage is low.

Attaching the bench result

<details><summary>Details</summary>

```
test Apple Vision OCR/Performance/ ... bench:   352973236 ns/iter (+/- 18703264)

Average Accuracy: 0.87
Average Accuracy: 0.87
Average Accuracy: 0.87

Warning: Unable to complete 10 samples in 10.0s. You may wish to increase target time to 18.5s or enable flat sampling.
Average Accuracy: 0.87
test Apple Vision OCR with Accuracy/Performance and Accuracy/ ... Average Accuracy: 0.87
Average Accuracy: 0.87
Average Accuracy: 0.87
Average Accuracy: 0.87
Average Accuracy: 0.87
Average Accuracy: 0.87
Average Accuracy: 0.87
Average Accuracy: 0.87
Average Accuracy: 0.87
Average Accuracy: 0.87
bench:   336888740 ns/iter (+/- 3900718)
```
</details> 

I think this is the best i could bring down the usage, please let me know if you have any ideas :)

**Comment by louis030195** [comment]
@b4s36t4 does lang correction improve accuracy?

also ideally we have a different way to measure CPU

i think just opening Instruments app with screenpipe CLI (audio off) with CPU counter and just see difference in terms of the shape

we want to avoid spikes in general because it prevents people to work with screenpipe in the background, while smooth CPU usage is fine

**Comment by b4s36t4** [comment]
Hi, @louis030195. I tried the instruments app to measure cpu usage (with sound). my observation is that cpu usage is high for ocr function is because grayscale converstion.

New Implementation:
Screenshot 2024-12-12 at 2 18 18 AM


On Main:
Screenshot 2024-12-12 at 2 20 53 AM

Not sure what could we do here, can you help?

**Comment by b4s36t4** [comment]
Also ran the screenpipe without sound for 3 sets of times (main branch, current commit, remove `grayscale()` function call).

Below are the results, I think it's as same as old one. 


```
After (remove grayscale function call):
2024-12-11T21:01:20.397320Z  INFO screenpipe_server::video: Starting FFmpeg process for file: /Users/mahesh/.screenpipe/data/monitor_1_2024-12-11_21-01-20.mp4    
2024-12-11T21:01:20.401430Z  WARN screenpipe_server::core: Failed to insert frame: error returned from database: (code: 517) database is locked    
2024-12-11T21:01:22.598606Z  INFO screenpipe_server::resource_monitor: Runtime: 10s, Total Memory: 3% (0 GB / 16 GB), Total CPU: 0%, NPU: N/A
2024-12-11T21:01:32.678886Z  INFO screenpipe_server::resource_monitor: Runtime: 20s, Total Memory: 3% (0 GB / 16 GB), Total CPU: 121%, NPU: N/A
2024-12-11T21:01:42.757563Z  INFO screenpipe_server::resource_monitor: Runtime: 30s, Total Memory: 4% (1 GB / 16 GB), Total CPU: 112%, NPU: N/A
2024-12-11T21:01:52.842521Z  INFO screenpipe_server::resource_monitor: Runtime: 40s, Total Memory: 4% (1 GB / 16 GB), Total CPU: 115%, NPU: N/A
2024-12-11T21:02:02.921543Z  INFO screenpipe_server::resource_monitor: Runtime: 50s, Total Memory: 4% (1 GB / 16 GB), Total CPU: 88%, NPU: N/A
2024-12-11T21:02:13.038761Z  INFO screenpipe_server::resource_monitor: Runtime: 60s, Total Memory: 4% (1 GB / 16 GB), Total CPU: 117%, NPU: N/A
2024-12-11T21:02:23.118254Z  INFO screenpipe_server::resource_monitor: Runtime: 70s, Total Memory: 5% (1 GB / 16 GB), Total CPU: 117%, NPU: N/A

Current Commit:
2024-12-11T21:03:19.686213Z  INFO screenpipe_server::video: Starting FFmpeg process for file: /Users/mahesh/.screenpipe/data/monitor_1_2024-12-11_21-03-19.mp4    
2024-12-11T21:03:20.197117Z  INFO screenpipe_server::resource_monitor: Runtime: 10s, Total Memory: 3% (0 GB / 16 GB), Total CPU: 0%, NPU: N/A
2024-12-11T21:03:30.278469Z  INFO screenpipe_server::resource_monitor: Runtime: 20s, Total Memory: 3% (1 GB / 16 GB), Total CPU: 117%, NPU: N/A
2024-12-11T21:03:40.483346Z  INFO screenpipe_server::resource_monitor: Runtime: 30s, Total Memory: 3% (1 GB / 16 GB), Total CPU: 116%, NPU: N/A
2024-12-11T21:03:50.563084Z  INFO screenpipe_server::resource_monitor: Runtime: 40s, Total Memory: 4% (1 GB / 16 GB), Total CPU: 87%, NPU: N/A
2024-12-11T21:04:00.642780Z  INFO screenpipe_server::resource_monitor: Runtime: 50s, Total Memory: 4% (1 GB / 16 GB), Total CPU: 122%, NPU: N/A
2024-12-11T21:04:10.722999Z  INFO screenpipe_server::resource_monitor: Runtime: 60s, Total Memory: 4% (1 GB / 16 GB), Total CPU: 116%, NPU: N/A
2024-12-11T21:04:20.803155Z  INFO screenpipe_server::resource_monitor: Runtime: 70s, Total Memory: 5% (1 GB / 16 GB), Total CPU: 90%, NPU: N/A
^C2024-12-11T21:04:23.381472Z  INFO screenpipe: received ctrl+c, initiating shutdown

Main Branch:
2024-12-11T21:05:15.442599Z  WARN screenpipe_server::core: Failed to insert frame: error returned from database: (code: 517) database is locked    
2024-12-11T21:05:16.817849Z  INFO screenpipe_server::resource_monitor: Runtime: 10s, Total Memory: 3% (1 GB / 16 GB), Total CPU: 0%, NPU: N/A
2024-12-11T21:05:26.898576Z  INFO screenpipe_server::resource_monitor: Runtime: 20s, Total Memory: 4% (1 GB / 16 GB), Total CPU: 120%, NPU: N/A
2024-12-11T21:05:36.974961Z  INFO screenpipe_server::resource_monitor: Runtime: 30s, Total Memory: 4% (1 GB / 16 GB), Total CPU: 111%, NPU: N/A
2024-12-11T21:05:47.057148Z  INFO screenpipe_server::resource_monitor: Runtime: 40s, Total Memory: 4% (1 GB / 16 GB), Total CPU: 90%, NPU: N/A
2024-12-11T21:05:57.137085Z  INFO screenpipe_server::resource_monitor: Runtime: 50s, Total Memory: 4% (1 GB / 16 GB), Total CPU: 121%, NPU: N/A
2024-12-11T21:06:07.218362Z  INFO screenpipe_server::resource_monitor: Runtime: 60s, Total Memory: 4% (1 GB / 16 GB), Total CPU: 113%, NPU: N/A
2024-12-11T21:06:17.300261Z  INFO screenpipe_server::resource_monitor: Runtime: 70s, Total Memory: 5% (1 GB / 16 GB), Total CPU: 114%, NPU: N/A
```

**Comment by b4s36t4** [comment]
I have a instrument recording of 1min frame. does that works or 30min should be the bench?

**Comment by louis030195** [comment]
> I have a instrument recording of 1min frame. does that works or 30min should be the bench?

sure

also i don't really trust the `resource_monitor` it does not seem accurate

**Comment by louis030195** [comment]
any update

**Comment by b4s36t4** [comment]
Sorry, missed it.

I have attached the traces I have ran for a minute in two (main and feature branch).

[native.trace.zip](https://github.com/user-attachments/files/18170203/native.trace.zip)

[Main Trace.trace.zip](https://github.com/user-attachments/files/18170193/Main.Trace.trace.zip)

if these are not good, let me know would record for more minutes.

**Comment by b4s36t4** [comment]
Hi, @louis030195 Sorry for the delay. But fixed the conflict, please check! Thanks :)

**Comment by louis030195** [comment]
/approve 

thank you for the amazing work and patience!

**Comment by louis030195** [comment]
@b4s36t4 i'm afraid there is a memory leak

memory increased to 16gb

when i turn off vision it does not happen

**Comment by louis030195** [comment]
can you help?

**Comment by b4s36t4** [comment]
Sure will do, but I need help on how to debug. Can you help me with reproduction or anything you have with you?

**Comment by louis030195** [comment]
try --fps 30

**Comment by louis030195** [comment]
i think i fixed it check last commits

**Comment by b4s36t4** [comment]
Nice, sorry for not able to help instantly. But why it is happening why do we need to wrap the login in `auto release` func!?

**Comment by louis030195** [comment]
idk there is a leak somewhere

**Comment by b4s36t4** [comment]
Got it, will take a deep look and see if still there are any leaks and will fix if there's any.

**Pull request merged**